content/courses/program-security/arbitrary-cpi.md

```diff
@@ -80,59 +79,60 @@ pub mod arbitrary_cpi_insecure {
+    authority: Signer<'info>,
     token_program: UncheckedAccount<'info>,
 }
 ```
 
 An attacker could easily call this instruction and pass in a duplicate token
 program that they created and control.
 
-### Add program checks
+### Add Program Checks
 
-It's possible to fix this vulnerabilty by simply adding a few lines to the `cpi`
-instruction to check whether or not `token_program`'s public key is that of the
-SPL Token Program.
+It's possible to fix this vulnerability by simply adding a few lines to the
+`cpi` instruction to check whether or not `token_program`'s public key is that
+of the SPL Token Program.
 
 ```rust
-pub fn cpi_secure(ctx: Context<Cpi>, amount: u64) -> ProgramResult {
-    if &spl_token::ID != ctx.accounts.token_program.key {
-        return Err(ProgramError::IncorrectProgramId);
+pub fn cpi_secure(ctx: Context<Cpi>, amount: u64) -> Result<()> {
+    if ctx.accounts.token_program.key() != spl_token::ID {
+        return Err(ProgramError::IncorrectProgramId.into());
     }
     solana_program::program::invoke(
         &spl_token::instruction::transfer(
-            ctx.accounts.token_program.key,
-            ctx.accounts.source.key,
-            ctx.accounts.destination.key,
-            ctx.accounts.authority.key,
+            ctx.accounts.token_program.key(),
+            ctx.accounts.source.key(),
+            ctx.accounts.destination.key(),
+            ctx.accounts.authority.key(),
             &[],
             amount,
         )?,
         &[
-            ctx.accounts.source.clone(),
-            ctx.accounts.destination.clone(),
-            ctx.accounts.authority.clone(),
+            ctx.accounts.source.to_account_info(),
+            ctx.accounts.destination.to_account_info(),
+            ctx.accounts.authority.to_account_info(),
         ],
-    )
+    )?;
+    Ok(())
 }
 ```
 
 Now, if an attacker passes in a different token program, the instruction will
 return the `ProgramError::IncorrectProgramId` error.
 
-Depending on the program you’re invoking with your CPI, you can either hard code
-the address of the expected program ID or use the program’s Rust crate to get
+Depending on the program you're invoking with your CPI, you can either hard code
+the address of the expected program ID or use the program's Rust crate to get
 the address of the program, if available. In the example above, the `spl_token`
 crate provides the address of the SPL Token Program.
 
-### Use an Anchor CPI module
+### Use an Anchor CPI Module
 
 A simpler way to manage program checks is to use Anchor CPI modules. We learned
-in a
-[previous lesson](https://github.com/Unboxed-Software/solana-course/blob/main/content/anchor-cpi)
-that Anchor can automatically generate CPI modules to make CPIs into the program
-simpler. These modules also enhance security by verifying the public key of the
-program that’s passed into one of its public instructions.
+in a [previous lesson](../anchor-cpi/README.md) that Anchor can automatically
```

Comment:
Use metadata.yaml to find the actual URL of the previous lesson and a make a link to it per CONTRIBUTING.

Reply:
Got it